<%
  import djerba.plugins.pwgs.constants as constants
  from djerba.util.html import html_builder
%>

<!-- ctDNA Detection -->
${html_builder.section_cells_begin("Mutational Sampling", True)}

      <img id='Detection' style='width: 100%; object-fit: contain' src="${results.get(constants.PWGS_PLOT)}"/>

      <p> This test counted the number of sites in the patient's plasma 
            at which variants were detected that match variants found in the patient's primary tumour 
            (candidate sites passing QC: <strong>${html_builder.k_comma_format(results.get(constants.SITES_CHECKED))}</strong>).
            To test for the possibility that candidate SNVs match SNVs arising from sequencing artefacts, 
            the number of detected sites in the plasma sample was statistically 
            compared to the number of detected sites in a healthy control cohort. 
            The number of SNVs of tumour origin 
            detected in this plasma sample (detected sites: <strong>${html_builder.k_comma_format(results.get(constants.SITES_DETECTED))}</strong>) was
             <strong>${results.get(constants.SIGNIFICANCE)} </strong>  
             than in the control cohort (N=${results.get(constants.COHORT_N)}).</p>
            <table class="variants" style="width:100%">
                  <thead>
                        <th style=" width:16%">Reads Checked</th>
                        <th style=" width:16%">Reads Detected</th>
                        <th style=" width:20%">Detected:Checked Reads Ratio (%)</th>
                  </thead>
                  <tbody>
                        <tr>
                              <td>${html_builder.k_comma_format(results.get(constants.READS_CHECKED))}</td>
                              <td>${html_builder.k_comma_format(results.get(constants.READS_DETECTED))}</td>
                              <td>${results.get(constants.TUMOUR_FRACTION_READS)}</td>
                        </tr>
                  </tbody>
            </table>

${html_builder.section_cells_end()}