
## This file is the HTML code for the Copy Number Variation (CNV) section of the report.

<%
  import djerba.plugins.cnv.constants as cnv
  from djerba.plugins.cnv.html import make_table_header, make_table_rows
  from djerba.util.html import html_builder
%>
  
<!-- Copy Number Variants -->

${html_builder.section_cells_begin("Copy Number Variation", True)}

      <p>The percent genome altered (PGA) was <strong>${results.get(cnv.PERCENT_GENOME_ALTERED)}</strong>%. 
            <strong>${results.get(cnv.TOTAL_VARIANTS)}</strong> cancer gene(s) were subject to copy number variation, 
            of which <strong>${results.get(cnv.CLINICALLY_RELEVANT_VARIANTS)}</strong> corresponded to an oncogenic alteration, as defined by OncoKB. 
	    Regions with large copy number gains (&#8805; 6 CN) marked as &#9650; in plot below.
      </p>

      <img id='CNV' style='width: 100%; object-fit: contain' src="${results.get('cnv plot')}"/>
      % if results.get(cnv.CLINICALLY_RELEVANT_VARIANTS) > 0:
            <table class="variants" width="100%">
            ${make_table_header()}
            <tbody>
            % for row in make_table_rows(results.get(cnv.BODY)):
            ${row}
            % endfor
            </table>
            </tbody>
            </table>
	        % if results.get(cnv.HAS_EXPRESSION_DATA):
                <table class="suppl" width="100%">
                  <tr><td><b>Expr. (%)</b>: Expression Percentile for gene mRNA, or NA if comparison data is not available</td></tr>
		</table>
	        % endif
	    </table>
      % endif
      
${html_builder.section_cells_end()}
