<p><strong>Assay description</strong></p>
<p>This test was developed and its performance characteristics determined by OICR Genomics. It has not been cleared or approved by the US Food and Drug Administration.</p>
<p>This assay is a comprehensive next generation sequencing assay, specifically DNA-based whole genome sequencing (WGS). WGS libraries are prepared using the KAPA Hyper Prep kit with DNA extracted from FFPE or fresh frozen tissue (for tumour samples) or buffy coat blood specimens (for matched normal blood samples). Paired-end sequencing is performed using the NovaSeq6000 technology. For whole genome sequencing, alignments are performed using bwa mem (0.7.12) against reference genome hg38 and processed according to GATK best practices, and include duplicate marking with Picard (2.21.2), indel realignment, and base recalibration with GATK (4.1.6.0). SNVs and INDELs are called using MuTect2 (GATK 4.2.6.1) and annotated with VariantEffectPredictor (105.0) followed by OncoKB. Copy number variations are called using Sequenza (2.1.2).</p>
<p>Variant prioritization is based on OncoKB actionable tiers (described below). In addition, variants without an OncoKB actionable tier are reported if they are listed as oncogenic, likely oncogenic or predicted oncogenic. Based on a minimum tumour purity of 30%, the sensitivity for SNVs and INDELs is 96% and 89%, respectively. The sensitivity for CNVs and FNA fusions is 86% and 32%, respectively. The limit of detection is 10% VAF for SNVs and 20% for INDELs.</p>
<p>Assay results are collated into the report document by <a href="https://github.com/oicr-gsi/djerba">Djerba</a> (${djerba_version}).</p>
<p><strong>Genes tested</strong></p>
<p>Although whole genome sequencing encompasses all genes in a specimen, this report is restricted to cancer genes as defined by OncoKB as of the date this report.</p>
