<%
  from time import strftime
  import djerba.render.constants as constants
  from djerba.render.json_to_html import html_builder
  from djerba.util.image_to_base64 import converter
%>

<!-- Genomic Landscape -->
${html_builder().section_cells_begin("<h2>Genomic Landscape</h2>","main")}
      
      <p >
            Tumour Mutation Burden (TMB) was <strong>${genomic_landscape_info.get(constants.TMB_PER_MB)}</strong> coding mutations per Mb (${genomic_landscape_info.get(constants.TMB_TOTAL)} mutations)
            which corresponded to the ${html_builder().make_ordinal(genomic_landscape_info.get(constants.PAN_CANCER_PERCENTILE))} 
            percentile of the pan-cancer cohort and classifies it as <strong>${html_builder().pull_biomarker_text(genomic_biomarkers,"TMB")}</strong>.       

            % if genomic_landscape_info.get(constants.CANCER_SPECIFIC_PERCENTILE) != "NA" :
                  This TMB places the tumour in the <strong>${html_builder().make_ordinal(genomic_landscape_info.get(constants.CANCER_SPECIFIC_PERCENTILE))}</strong> percentile of the ${genomic_landscape_info.get(constants.CANCER_SPECIFIC_COHORT)} cohort.
            % endif

            % if sample_info_and_quality.get(constants.PURITY_PERCENT) > 50:
                  The microsatellite status is <strong>${html_builder().pull_biomarker_text(genomic_biomarkers,"MSI")} </strong>.
            % endif 
      </p>

      <table><tr>
            <td>
                  <img id='TMB' style='width: 100%; object-fit: contain' src="${converter().convert_svg(tmb_plot, 'TMB plot')}"/>
            </td>
            <td>
                  <img id='PGA' style='width: 100%; object-fit: contain' src="${converter().convert_svg(pga_plot, 'PGA plot')}"/>
            </td>
      </tr></table>

      <!-- other biomarkers table -->
      % if sample_info_and_quality.get(constants.PURITY_PERCENT) > 50:
            <table class="variants" style="width:100%">
                  <thead>
                        <th style=" width:10%">Biomarker</th>
                        <th style=" width:10%">Call</th>
                        <th style=" width:80%">Score & Confidence</th>
                  </thead>
                  <tbody>
                        % for row in html_builder().biomarker_table_rows(genomic_biomarkers):
                        ${row}
                        % endfor 
                  </tbody>
            </table>
      % endif 


${html_builder().section_cells_end()}

${html_builder().section_cells_begin("<h2>SNVs and in/dels</h2>","main")}
   
      <p>
            <strong>${html_builder().k_comma_format(small_mutations_and_indels.get(constants.TOTAL_VARIANTS))}</strong> somatic mutation(s) were detected in exonic or splice regions, 
            of which <strong>${genomic_landscape_info.get(constants.TMB_TOTAL)}</strong> impacted a coding sequence,
            and <strong>${small_mutations_and_indels.get(constants.CLINICALLY_RELEVANT_VARIANTS)}</strong> corresponded to an oncogenic mutation, as defined by OncoKB.
      </p>

      <img id='VAF' style='width: 100%; object-fit: contain' src="${converter().convert_svg(vaf_plot, 'VAF plot')}"/>
      
      % if small_mutations_and_indels.get(constants.CLINICALLY_RELEVANT_VARIANTS) > 0:
            <table class="variants" width="100%">
                  ${html_builder().oncogenic_small_mutations_and_indels_header(small_mutations_and_indels)}
                  <tbody>
                  % for row in html_builder().oncogenic_small_mutations_and_indels_rows(small_mutations_and_indels):
                        ${row}
                  % endfor
                  </tbody>
            </table>
	    <table class="suppl" width="100%">
              <tr>
		<td width="50%" vertical-align="top"><b>Chr.</b>: Chromosome and cytoband</td>
		% if small_mutations_and_indels.get(constants.HAS_EXPRESSION_DATA):
		<td width="50%" vertical-align="top"><b>Expr. (%)</b>: Expression Percentile for gene mRNA, or NA if comparison data is not available</td></tr>
	        % endif
	    </table>
      % endif

      
${html_builder().section_cells_end()}

<!-- Copy Number Variants -->
${html_builder().section_cells_begin("<h2>Copy Number Variation</h2>","main")}

      <p>The percent genome altered (PGA) was <strong>${genomic_landscape_info.get(constants.PERCENT_GENOME_ALTERED)}</strong>%. 
            <strong>${oncogenic_somatic_CNVs.get(constants.TOTAL_VARIANTS)}</strong> cancer gene(s) were subject to copy number variation, 
            of which <strong>${oncogenic_somatic_CNVs.get(constants.CLINICALLY_RELEVANT_VARIANTS)}</strong> corresponded to an oncogenic alteration, as defined by OncoKB. 
            Small regions (&#60;3mb) with large copy number gains marked as &#9650 in plot below.
      </p>

      <img id='CNV' style='width: 100%; object-fit: contain' src="${converter().convert_svg(cnv_plot, 'CNV plot')}"/>

      % if oncogenic_somatic_CNVs.get(constants.CLINICALLY_RELEVANT_VARIANTS) > 0:
            <table class="variants" width="100%">
            ${html_builder().oncogenic_CNVs_header(oncogenic_somatic_CNVs)}
            <tbody>
            % for row in html_builder().oncogenic_CNVs_rows(oncogenic_somatic_CNVs):
                  ${row}
            % endfor
            </table>
            </tbody>
            </table>
	        % if oncogenic_somatic_CNVs.get(constants.HAS_EXPRESSION_DATA):
                <table class="suppl" width="100%">
                  <tr><td><b>Expr. (%)</b>: Expression Percentile for gene mRNA, or NA if comparison data is not available</td></tr>
		</table>
	        % endif
	    </table>

      % endif
      
${html_builder().section_cells_end()}

% if assay_type == 'WGTS':
      ${html_builder().section_cells_begin("<h2>Fusions and Structural Variants</h2>","main")}

            <p><strong>${structural_variants_and_fusions.get(constants.TOTAL_VARIANTS)}</strong> cancer gene(s) were subject to rearrangement, of which <strong>${structural_variants_and_fusions.get(constants.CLINICALLY_RELEVANT_VARIANTS)}</strong> corresponded to oncogenic fusions, as defined by OncoKB.</p>

            % if structural_variants_and_fusions.get(constants.CLINICALLY_RELEVANT_VARIANTS) > 0:
                  <table class="variants" width="100%">
                  ${html_builder().structural_variants_and_fusions_header()}
                  <tbody>
                  % for row in html_builder().structural_variants_and_fusions_rows(structural_variants_and_fusions):
                        ${row}
                  % endfor
                  </tbody>
                  </table>
            % endif
            
      ${html_builder().section_cells_end()}
% endif
