<%
  from time import strftime
  import djerba.render.constants as constants
  from djerba.render.json_to_html import html_builder
  from djerba.util.image_to_base64 import converter
%>

<div class="keep-together">
<div style="height:36px"><h3 class="header1">Genomic Landscape</h3></div>
<table width="100%">
% for row in html_builder().genomic_landscape_table_rows(genomic_landscape_info):
      ${row}
% endfor
</table>
## ends	keep-together
</div>
## image sources may be file paths (for testing) or base64 data blobs (for a self-contained document)
<div>
  <p><img style='width: 50%; object-fit: contain' src="${converter().convert_svg(tmb_plot, 'TMB plot')}"/><img style='width: 50%; object-fit: contain' src="${converter().convert_svg(vaf_plot, 'VAF plot')}"/></p>
</div>
<div class="keep-together">
<div style="height:36px">
<h3 class="header1">Oncogenic Small Mutations and Indels</h3>
</div>
<p><strong>${small_mutations_and_indels.get(constants.TOTAL_VARIANTS)}</strong> somatic mutations were detected, of which <strong>${small_mutations_and_indels.get(constants.CLINICALLY_RELEVANT_VARIANTS)}</strong> corresponded to an oncogenic mutation, as defined by OncoKB.</p>
% if small_mutations_and_indels.get(constants.CLINICALLY_RELEVANT_VARIANTS) > 0:
<table class="variants" width="100%">
${html_builder().oncogenic_small_mutations_and_indels_header()}
<tbody>
% for row in html_builder().oncogenic_small_mutations_and_indels_rows(small_mutations_and_indels):
      ${row}
% endfor
</tbody>
</table>
% endif
## ends	keep-together
</div>
<div class="keep-together">
<div style="height:36px">
<h3 class="header1">Oncogenic Somatic Copy Number Variation</h3>
</div>
<p><strong>${oncogenic_somatic_CNVs.get(constants.TOTAL_VARIANTS)}</strong> cancer genes were subject to copy number variation, of which <strong>${oncogenic_somatic_CNVs.get(constants.CLINICALLY_RELEVANT_VARIANTS)}</strong> corresponded to an oncogenic alteration, as defined by OncoKB.</p>
% if oncogenic_somatic_CNVs.get(constants.CLINICALLY_RELEVANT_VARIANTS) > 0:
<table class="variants" width="100%">
${html_builder().oncogenic_CNVs_header()}
<tbody>
% for row in html_builder().oncogenic_CNVs_rows(oncogenic_somatic_CNVs):
      ${row}
% endfor
</table>
</tbody>
</table>
% endif
## ends	keep-together
</div>
% if assay_type == 'WGTS':
<div class="keep-together">
<div style="height:36px">
<h3 class="header1">Structural Variants and Fusions</h3>
</div>
<p><strong>${structural_variants_and_fusions.get(constants.TOTAL_VARIANTS)}</strong> cancer genes were subject to rearrangement, of which <strong>${structural_variants_and_fusions.get(constants.CLINICALLY_RELEVANT_VARIANTS)}</strong> corresponded to oncogenic fusions, as defined by OncoKB.</p>
% if structural_variants_and_fusions.get(constants.CLINICALLY_RELEVANT_VARIANTS) > 0:
<table class="variants" width="100%">
${html_builder().structural_variants_and_fusions_header()}
<tbody>
% for row in html_builder().structural_variants_and_fusions_rows(structural_variants_and_fusions):
      ${row}
% endfor
</tbody>
</table>
% endif
## ends	keep-together
</div>
% endif
<div class="keep-together">
<div style="height:36px">
<h3 class="header1">Supplementary Gene Information</h3>
</div>
% if len(gene_info) > 0:
<table class="variants" width="100%">
${html_builder().supplementary_gene_info_header()}
<tbody>
% for row in html_builder().supplementary_gene_info_rows(gene_info):
      ${row}
% endfor
</tbody>
</table>
% else:
<p style="font-size:14px">No relevant oncogenes were detected in OncoKB.</p>
## ends	keep-together
</div>
% endif
