<%
  from time import strftime
  import djerba.render.constants as constants
  from djerba.render.json_to_html import html_builder
  from djerba.util.image_to_base64 import converter
%>

<!-- Genomic Landscape -->
${html_builder().make_sections_into_cells("<h2>Genomic Landscape</h2>","main")}
      
      <p >
            Tumour Mutation Burden (TMB) was <strong>${genomic_landscape_info.get(constants.TMB_PER_MB)}</strong> coding mutations per Mb (${genomic_landscape_info.get(constants.TMB_TOTAL)} mutations)
            which corresponded to the <strong>${html_builder().make_ordinal(${genomic_landscape_info.get(constants.PAN_CANCER_PERCENTILE)})}</strong> 
            percentile of the pan-cancer cohort and classifies it as <strong>${html_builder().pull_biomarker_text(genomic_biomarkers,"TMB")} </strong>. 
      </p>

      <table><tr>
            <td>
                  <img id='TMB' style='width: 100%; object-fit: contain' src="${converter().convert_svg(tmb_plot, 'TMB plot')}"/>
            </td>
            <td>
                  <img id='PGA' style='width: 100%; object-fit: contain' src="${converter().convert_svg(pga_plot, 'PGA plot')}"/>
            </td>
      </tr></table>

</div></div>

${html_builder().make_sections_into_cells("<h2>Small Mutations</h2>","main")}
   
      <p>
            <strong>${html_builder().k_comma_format(${small_mutations_and_indels.get(constants.TOTAL_VARIANTS)})</strong> somatic mutation(s) were detected, 
            of which <strong>${genomic_landscape_info.get(constants.TMB_TOTAL)}</strong> impacted a coding sequence,
            and <strong>${small_mutations_and_indels.get(constants.CLINICALLY_RELEVANT_VARIANTS)}</strong> corresponded to an oncogenic mutation, as defined by OncoKB.
      </p>

      <img id='VAF' style='width: 100%; object-fit: contain' src="${converter().convert_svg(vaf_plot, 'VAF plot')}"/>
      
      % if small_mutations_and_indels.get(constants.CLINICALLY_RELEVANT_VARIANTS) > 0:
            <table class="variants" width="100%">
                  ${html_builder().oncogenic_small_mutations_and_indels_header()}
                  <tbody>
                  % for row in html_builder().oncogenic_small_mutations_and_indels_rows(small_mutations_and_indels):
                        ${row}
                  % endfor
                  </tbody>
            </table>
      % endif

      
</div></div>

<!-- Copy Number Variants -->
${html_builder().make_sections_into_cells("<h2>Copy Number Variation</h2>","main")}

      <p>The percent genome altered (PGA) was <strong>${genomic_landscape_info.get(constants.PERCENT_GENOME_ALTERED)}</strong>%. 
            <strong>${oncogenic_somatic_CNVs.get(constants.TOTAL_VARIANTS)}</strong> cancer gene(s) were subject to copy number variation, 
            of which <strong>${oncogenic_somatic_CNVs.get(constants.CLINICALLY_RELEVANT_VARIANTS)}</strong> corresponded to an oncogenic alteration, as defined by OncoKB. 
            Amplifications greater than four copies marked shown as &#9650 in plot below.
      </p>

      <img id='CNV' style='width: 100%; object-fit: contain' src="${converter().convert_svg(cnv_plot, 'CNV plot')}"/>

      % if oncogenic_somatic_CNVs.get(constants.CLINICALLY_RELEVANT_VARIANTS) > 0:
            <table class="variants" width="100%">
            ${html_builder().oncogenic_CNVs_header()}
            <tbody>
            % for row in html_builder().oncogenic_CNVs_rows(oncogenic_somatic_CNVs):
                  ${row}
            % endfor
            </table>
            </tbody>
            </table>
      % endif
      
</div></div>

% if assay_type == 'WGTS':
      ${html_builder().make_sections_into_cells("<h2>Fusions and Structural Variants</h2>","main")}

            <p><strong>${structural_variants_and_fusions.get(constants.TOTAL_VARIANTS)}</strong> cancer gene(s) were subject to rearrangement, of which <strong>${structural_variants_and_fusions.get(constants.CLINICALLY_RELEVANT_VARIANTS)}</strong> corresponded to oncogenic fusions, as defined by OncoKB.</p>

            % if structural_variants_and_fusions.get(constants.CLINICALLY_RELEVANT_VARIANTS) > 0:
                  <table class="variants" width="100%">
                  ${html_builder().structural_variants_and_fusions_header()}
                  <tbody>
                  % for row in html_builder().structural_variants_and_fusions_rows(structural_variants_and_fusions):
                        ${row}
                  % endfor
                  </tbody>
                  </table>
            % endif
            
      </div></div>
% endif
