<%
  from time import strftime
  import djerba.render.constants as constants
  from djerba.render.json_to_html import html_builder
  from djerba.util.image_to_base64 import converter
%>

<!-- Genomic Landscape -->
<div class="section">
      <div class="section-bar"><hr class="big-line"  ></div>

      <div class="section-title" >
        <h2>Genomic Landscape</h2>
      </div>

      <div class="section-content" >
            <p >Tumour Mutation Burden is <strong>${genomic_landscape_info.get(constants.TMB_TOTAL)}</strong> mutations (${genomic_landscape_info.get(constants.TMB_PER_MB)} coding mutations per Mb)
                   which corresponds to the <strong>${genomic_landscape_info.get(constants.PAN_CANCER_PERCENTILE)}</strong>th 
                  percentile of the pan-cancer cohort (${genomic_landscape_info.get(constants.PAN_CANCER_COHORT)}). The percent genome altered (PGA) is <strong>${genomic_landscape_info.get(constants.PERCENT_GENOME_ALTERED)}</strong>%. 
                  The tumour is <strong>
                        % for row in html_builder().msi_table_rows(genomic_biomarkers,"TMB"):
                        ${row}
                        % endfor
                        </strong>. The microsatellite status is <strong>
                              % for row in html_builder().msi_table_rows(genomic_biomarkers,"MSI"):
                              ${row}
                              % endfor
                  </strong>. 
                </p>

    
            <table><tr>
                  <td>
                        <img id='TMB' style='width: 100%; object-fit: contain' src="${converter().convert_svg(tmb_plot, 'TMB plot')}"/>
                  </td>
                  <td>
                        <img id='PGA' style='width: 100%; object-fit: contain' src="${converter().convert_svg(pga_plot, 'PGA plot')}"/>
                  </td>
            </tr></table>

            <table class="variants" style="width:100%">
                  <thead>
                    <th style=" width:25%">Biomarker</th>
                    <th style=" width:10%">Call</th>
                    <th style=" width:100%">Score & Confidence</th>
                  </thead>
                  <tbody>
                    <tr>
                      <td>Microsatellite Instability</td>
                      <td>
                        % for row in html_builder().msi_table_rows(genomic_biomarkers,"MSI"):
                        ${row}
                        % endfor
                        </td>
                      <td>
                      <img id='MSI' style='width: 100%; mix-blend-mode: multiply;' src="${converter().convert_svg(msi_plot, 'MSI plot')}
                      "/>  
                      </td>
                    </tr>
                  </tbody>
            </table>
    
            ## 
      </div>
</div>
    
<div class="section">
      <div class="section-bar" ><hr class="big-line" ></div>

      <div class="section-title" >
        <h2 >Small Mutations</h2>
      </div>
      
      <div class="section-content" >

            <p><strong>${small_mutations_and_indels.get(constants.TOTAL_VARIANTS)}</strong> somatic mutations were detected, of which <strong>${small_mutations_and_indels.get(constants.CLINICALLY_RELEVANT_VARIANTS)}</strong> corresponded to an oncogenic mutation, as defined by OncoKB.</p>

            <img id='VAF' style='width: 100%; object-fit: contain' src="${converter().convert_svg(vaf_plot, 'VAF plot')}"/>
            
            % if small_mutations_and_indels.get(constants.CLINICALLY_RELEVANT_VARIANTS) > 0:
                  <table class="variants" width="100%">
                        ${html_builder().oncogenic_small_mutations_and_indels_header()}
                        <tbody>
                        % for row in html_builder().oncogenic_small_mutations_and_indels_rows(small_mutations_and_indels):
                              ${row}
                        % endfor
                        </tbody>
                  </table>
            % endif

      </div>
</div>

    <div class="section">
      <div class="section-bar" ><hr class="big-line" ></div>

      <div class="section-title" >
        <h2 >Copy Number <br>Variation</h2>
      </div>

      <div class="section-content" >

            <p><strong>${oncogenic_somatic_CNVs.get(constants.TOTAL_VARIANTS)}</strong> cancer genes were subject to copy number variation, of which <strong>${oncogenic_somatic_CNVs.get(constants.CLINICALLY_RELEVANT_VARIANTS)}</strong> corresponded to an oncogenic alteration, as defined by OncoKB.</p>

            <img id='CNV' style='width: 100%; object-fit: contain' src="${converter().convert_svg(cnv_plot, 'CNV plot')}"/>

            % if oncogenic_somatic_CNVs.get(constants.CLINICALLY_RELEVANT_VARIANTS) > 0:
                  <table class="variants" width="100%">
                  ${html_builder().oncogenic_CNVs_header()}
                  <tbody>
                  % for row in html_builder().oncogenic_CNVs_rows(oncogenic_somatic_CNVs):
                        ${row}
                  % endfor
                  </table>
                  </tbody>
                  </table>
            % endif
      </div>
</div>

% if assay_type == 'WGTS':
      <div class="section">
            <div class="section-bar" ><hr class="big-line" ></div>
            <div class="section-title" >
                  <h2 >Fusions and <br> Structural Variants</h2>
            </div>

            <div class="section-content">

                  <p><strong>${structural_variants_and_fusions.get(constants.TOTAL_VARIANTS)}</strong> cancer genes were subject to rearrangement, of which <strong>${structural_variants_and_fusions.get(constants.CLINICALLY_RELEVANT_VARIANTS)}</strong> corresponded to oncogenic fusions, as defined by OncoKB.</p>
            
                  % if structural_variants_and_fusions.get(constants.CLINICALLY_RELEVANT_VARIANTS) > 0:
                        <table class="variants" width="100%">
                        ${html_builder().structural_variants_and_fusions_header()}
                        <tbody>
                        % for row in html_builder().structural_variants_and_fusions_rows(structural_variants_and_fusions):
                              ${row}
                        % endfor
                        </tbody>
                        </table>
                  % endif
            </div>
      </div>
% endif