<%
  import djerba.render.constants as constants
  from djerba.render.json_to_html import html_builder
%>

  <hr class="header-big-line" style="width:100%; align:right; page-break-before:always;">
  <h2>Supplementary </h2>

  % if not failed:
    ${html_builder().section_cells_begin("<h3>Gene Information</h3>","supp")}

    % if len(gene_info) > 0:
          <table class="variants" width="100%" >
          ${html_builder().supplementary_gene_info_header()}
          <tbody>
          % for row in html_builder().supplementary_gene_info_rows(gene_info):
                ${row}
          % endfor
          </tbody>
          </table>
    % else:
          <p >No relevant oncogenes were detected in OncoKB.</p>

    % endif
    <p><strong>Genes tested</strong>: Although whole genome sequencing encompasses all genes in a specimen, this report is restricted to cancer genes as defined by OncoKB as of the date this report.</p>

    ${html_builder().section_cells_end()}

    ${html_builder().section_cells_begin("<h3>OncoKB Definitions</h3>","supp")}
      <%include file="templates_for_supp/definitions_oncokb.html"/>
    ${html_builder().section_cells_end()}

  % endif

  ${html_builder().section_cells_begin("<h3>Definitions</h3>","supp")}

    <%include file="templates_for_supp/definitions_metrics.html"/>
    % if not failed:
      <%include file="templates_for_supp/definitions_tests.html"/>
    % endif

  ${html_builder().section_cells_end()}

  ${html_builder().section_cells_begin("<h3>Assay description</h3>","supp")}

    % if assay_type == constants.ASSAY_WGTS:
      <p> This assay combines two comprehensive next generation sequencing assays: a DNA-based whole genome sequencing (WGS) assay and an RNA-based whole transcriptome sequencing (WTS) assay.</p>
      <p>
          Whole Genome libraries were prepared using the KAPA Hyper Prep kit with DNA extracted from FFPE or fresh frozen tissue (for tumour samples) 
          or buffy coat blood specimens (for matched normal blood samples). Paired-end sequencing was performed using the Illumina NovaSeq6000 technology. 
          Reads were aligned using <a href="https://bio-bwa.sourceforge.net/">bwa</a> mem (${config.versions.get(constants.BWAMEM_VERSION)}) against
          reference genome <a href="https://www.ncbi.nlm.nih.gov/assembly/GCF_000001405.38/">GRCh38.p12</a> and processed according to GATK best practices, 
          including duplicate marking with <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037052812-MarkDuplicates-Picard-">Picard</a> (${constants.PICARD_VERSION}), 
          realignment around small insertions and deletions (in/dels), and base recalibration with <a href="https://gatk.broadinstitute.org/hc/en-us/sections/360008481611-4-1-6-0">GATK</a> (${constants.GATK_VERSION}). 
          SNVs and in/dels were called using <a href="https://gatk.broadinstitute.org/hc/en-us/articles/5358911630107-Mutect2">MuTect2</a> (GATK ${constants.MUTECT2_VERSION}) 
          and annotated with <a href="https://useast.ensembl.org/info/docs/tools/vep/index.html">VariantEffectPredictor</a> (${constants.VARIANTEFFECTPREDICTOR_VERSION}) using MANE  transcripts (MANE Clinical version 1.0 when available, <a href="https://www.ncbi.nlm.nih.gov/refseq/MANE/#Select">MANE Select</a> version 1.0 for all other transcripts).
          Variants were further annotated for oncogenicity and actionability by <a href="https://api.oncokb.org/oncokb-annotator/gene-annotation">OncoKB</a>. 
          In cases where OncoKB does not use MANE Select, links in annotation use the corresponding alteration in OncoKB.
          Copy number variations were called using <a href="https://doi.org/10.1093/annonc/mdu479">Sequenza</a> (${constants.SEQUENZA_VERSION}). 
      </p>

    % elif assay_type == constants.ASSAY_WGS:
      <p>This assay is a comprehensive next generation sequencing assay, specifically DNA-based whole genome sequencing (WGS). </p>
      <%include file="templates_for_supp/description_wgs.html"/>
    % else:
      <% raise RuntimeError("Unknown assay type") %>
    % endif

    ##<p>Assay results were collated into the report document by <a href="https://github.com/oicr-gsi/djerba">Djerba</a> (${djerba_version}) using pipeline v.${pipeline_version}.</p>

  ${html_builder().section_cells_end()}

  ## Slightly hacky method to omit technical notes with null content
  % if not technical_notes.strip() == 'No technical notes supplied.':
    ${html_builder().section_cells_begin("<h3>Technical Notes</h3>","supp")}
    ${html_builder().markdown_to_html(technical_notes)}
    ${html_builder().section_cells_end()}
  % endif

  ${html_builder().section_cells_begin("<h3>Disclaimer</h3>","supp")}
  <%include file="templates_for_supp/disclaimer.html"/>

  ${html_builder().section_cells_end()}
  
