
<p>
  <br><strong>Minimal Residual Disease (MRD): </strong>malignant cells that share genetics with the original tumour cells but that remain in a patient who achieves remission
   (<em>see</em> DOI: <a href="https://pubmed.ncbi.nlm.nih.gov/29376520/">10.1038/nrc.2017.125</a>). 
  Detection of minimal residual disease is not conclusively associated with prognosis.  
  Treatment choices should be made on a case-by-case basis with the treating physician.

  <br><strong>circulating tumour DNA (ctDNA): </strong>DNA that comes from dying cancerous cells and is found in the bloodstream. Here, a marker for MRD. 

  <br><strong>Coverage: </strong>the mean number of bases covering each sequenced base. Minimum coverage threshold is 30x (with a target of 40x).

  <br><strong>Insert Size: </strong>the length in base pairs of the DNA between adapters, affected by both biology and laboratory. 

  <br><strong>Mutational Sampling Analysis: </strong>sampling and sequencing cell-free DNA in plasma 
  to search for single nucleotide variants matching the primary tumour. 

  <br><strong>Candidate SNVs: </strong>the number of single nucleotide variants from the primary tumour that can be used for ctDNA detection. 
  If the number of candidate SNVs is too low (below 4,000), mutational sampling analysis is unreliable.
  
  <br><strong>Detected-Checked Reads Ratio: </strong>the percentage of reads with primary tumour matches divided by the total checked reads covering candidate sites.
  When ctDNA is detected, it may be an estimate of the fraction of ctDNA in the blood, but if ctDNA is not detected, it is an estimate of sequencing artefact matches.
</p>
