<%
  import djerba.plugins.supplement.body.versioning as versions 
  import djerba.core.constants as core_constants
%>

<p> 
This assay combines two comprehensive next generation sequencing assays: a targeted sequencing (TAR) assay and a shallow whole genome sequencing (sWGS) assay.

Shallow whole genome sequencing (sWGS) libraries were prepared using the KAPA Hyper Prep kit with DNA extracted from FFPE, cfDNA, fresh frozen tissue (for tumour samples). 
Paired-end sequencing was performed using the ${versions.TAR_ILLUMINA_VERSION} illumina technology to a minimum target coverage of 0.1x. 
Alignments were performed using <a href=${versions.BWAMEM_LINK}>bwa mem</a> (${versions.BWAMEM1_VERSION})
against reference genome <a href=${versions.REFERENCE_GENOME_LINK}>${versions.REFERENCE_GENOME_VERSION}</a>.
Copy number amplifications were called using <a href=${versions.ICHORCNA_LINK}>ichorCNA</a>.

TAR libraries were prepared using the KAPA Hyper Prep kit with DNA extracted from FFPE, cfDNA, fresh frozen tissue (for tumour samples) or buffy coat blood specimens (for matched normal blood samples). 
Paired-end sequencing was performed using the ${versions.TAR_ILLUMINA_VERSION} illumina technology. 
Alignments were performed using the <a href=${versions.CONSENSUSCRUNCHER_LINK}>ConsensusCruncher</a> workflow running <a href=${versions.BWAMEM_LINK}>bwa mem</a> (${versions.BWAMEM1_VERSION})
against reference genome <a href=${versions.REFERENCE_GENOME_LINK}>${versions.REFERENCE_GENOME_VERSION}</a>. 
Unique molecule based error suppression was completed with <a href=${versions.CONSENSUSCRUNCHER_LINK}>ConsensusCruncher</a>. 
SNVs and INDELs were called using <a href=${versions.MUTECT2_LINK}>MuTect2</a> (GATK ${versions.MUTECT2_VERSION}).
Fourteen genes were annotated with <a href=${versions.VARIANTEFFECTPREDICTOR_LINK}>VariantEffectPredictor</a> (${versions.VARIANTEFFECTPREDICTOR_VERSION}) 
using transcipts from <a href=${versions.MANE_LINK}>MANE Select</a> version ${versions.MANE_VERSION}.
Variants were further annotated for oncogenicity and actionability by <a href=${versions.ONCOKB_LINK}>OncoKB</a>.  
Variants without an OncoKB actionable tier were reported if they were listed as oncogenic, likely oncogenic or predicted oncogenic. 
</p>
