  
<p><strong>Tumor Mutation Burden (TMB): </strong> 
  The number of somatic, coding, non-synonymous base substitutions 
  and short insertions and deletions (indels) per megabase of tumour genome.
<br>
<strong>Microsatellite stability score (MSI)</strong> 
represents the percentage of the genome's short repeats (microsatellites) with insertions or deletions in the tumour. 
Instability in microsatellite repeat regions is often caused by genetic or epigenetic alterations to genes in the mismatch 
repair (MMR) pathway (including <em>MLH1</em>, <em>MSH2</em>, and <em>MSH6</em>). Confidence intervals are based on 100 
sets of 500 randomly sampled microsatellite sites. Tumours with an MSI score greater than 15.0% are considered 
microsatellite unstable (MSI).
<br>
<strong></strong>Homologous Recombination Deficiency (HRD)</strong> 
is the loss of a key DNA damage repair pathway in cancer, 
canonically associated with loss-of-function mutations in <em>BRCA1/2</em>. 
HRD manifests as a composite genomic signature that includes 
somatic mutations (SNVs and in/dels), Loss-Of-Heterozygosity (LOH), copy number alterations 
and other structural variants, summed as an HRD score.
<br>
<strong>Percent Genome Altered (PGA): </strong>
The fraction of the genome with evidence of copy number change, for a given sample. 
<br>
<strong>Cohort: </strong>
If available, the name of an external dataset used to calculate TMB percentile. 
Percentiles for tumour mutation burden (TMB) are plotted and reported against the corresponding cohort(s).
<br>
<strong>TCGA Percentile: </strong>
Percentile of TMB scores are plotted relative to 
<a href="https://gdc.cancer.gov/about-data/publications/pancanatlas">all TCGA samples</a> or closest cohort.
<br>
<strong>Loss of Heterozygosity: </strong>
A genetic event where there is loss of allelic variation for a given gene, ex. by loss of a wild-type allele. The calculation for LOH is a modification of the calculation described in the <a href="https://github.com/hartwigmedical/hmftools/blob/master/purple/README.md">PURPLE documentation</a>. First, the variant copy number (VCN) is computed by adjusting the variant allele frequency (VAF) of a particular alteration by the purity and multiplying by the copy number (CN) for that gene. Then, based on the minor allele copy number (MACN) for that gene, the logic for the classification of LOH is as follows: 
<p style="text-indent:30px;">If MACN &#x2264; 0.5 and VCN &gt; CN-0.5, LOH = True.</p>
<p style="text-indent:30px;">If MACN &#x2264; 0.5 and VCN &lt; CN-0.5, LOH = False.</p>
<p style="text-indent:30px;">If MACN &gt; 0.5, LOH = False.</p>
<p><strong>Copy Number Variants:</strong>
Calls at the gene-scale represent levels derived from copy-number analysis as used for annotation by OncoKB: 
</p>
<table class="suppl" width="100%">
  <tr><td>Deletion</td><td>a deep loss, possibly a homozygous deletion, defined by less than 0.5 copies</td></tr>
  <tr><td>Amplification</td><td>a high-level amplification (more copies, often focal), defined as greater than 2.4Xploidy copies</td></tr>
</table>
