  
<p><strong>Tumor Mutation Burden (TMB): </strong> 
  The number of somatic, coding, non-synonymous base substitutions 
  and short insertions and deletions (indels) per megabase of tumour genome.
<br>
<strong>Microsatellite stability score (MSI)</strong> 
represents the percentage of the genome's short repeats (microsatellites) with insertions or deletions in the tumour. 
Instability in microsatellite repeat regions is often caused by genetic or epigenetic alterations to genes in the mismatch 
repair (MMR) pathway (including <em>MLH1</em>, <em>MSH2</em>, and <em>MSH6</em>). Confidence intervals are based on 100 
sets of 500 randomly sampled microsatellite sites. Tumours with an MSI score greater than 15.0% are considered 
microsatellite unstable (MSI).
<br>
<strong>Percent Genome Altered (PGA): </strong>
The fraction of the genome with evidence of copy number change, for a given sample. 
PGA is calculated as the fraction of the genome where the absolute value of log<sub>2</sub> 
of the tumor / normal depth ratio is &#8805 0.2 , using copy number segment data produced by Sequenza.
<br>
<strong>Cohort: </strong>
If available, the name of an external dataset used to calculate TMB percentile. 
Percentiles for tumour mutation burden (TMB) are plotted and reported against the corresponding cohort(s).
<br>
<strong>TCGA Percentile: </strong>
Percentile of TMB scores are plotted relative to 
<a href="https://gdc.cancer.gov/about-data/publications/pancanatlas">all TCGA samples</a> or closest cohort.
<br>
<strong>Copy State:</strong> 
Copy state changes were called according to the log<sub>2</sub> coverage-ratio, 
which can be used to calculate an approximate minimum copy number for a diploid tumour with 100% cancer cell content. 
As the log<sub>2</sub> coverage-ratio thresholds scale with tumour purity, absolute copy number may be lower than predicted, 
in tumours with lower cancer cell content. Calls represent levels derived from copy-number analysis, 
are used for annotation by OncoKB, and indicate the copy-number level per gene, as follows: 
</p>
<table class="suppl" width="100%">
  <tr><td>Deep Deletion</td><td>a deep loss, possibly a homozygous deletion</td></tr>
  <tr><td>Shallow Deletion</td><td>a shallow loss, possibly a heterozygous deletion</td></tr>
  <tr><td>Gain</td><td>a low-level gain (a few additional copies, often broad)</td></tr>
  <tr><td>Amplification</td><td>a high-level amplification (more copies, often focal)</td></tr>
</table>

