<p><strong>Assay description</strong></p>
<p>This test was developed and its performance characteristics determined by OICR Genomics. It has not been cleared or approved by the US Food and Drug Administration.</p>
<p> This assay combines two comprehensive next generation sequencing assays: a DNA-based whole genome sequencing (WGS) assay and an RNA-based whole transcriptome sequencing (WTS) assay. WGS libraries are prepared using the KAPA Hyper Prep kit with DNA extracted from FFPE or fresh frozen tissue (for tumour samples) or buffy coat blood specimens (for matched normal blood samples). Paired-end sequencing is performed using the NovaSeq6000 technology. For whole genome sequencing, alignments are performed using bwa mem (0.7.12) against reference genome hg38 and processed according to GATK best practices and includes duplicate marking with Picard (2.21.2), indel realignment, and base recalibration with GATK (4.1.6.0). SNVs and INDELs are called using MuTect2 (GATK 4.1.1.0) and annotated with VariantEffectPredictor (92.0) followed by OncoKB. Copy number variations are called using Sequenza (2.1.2). Structural alterations are called using Delly (0.8.1) and post-processed using MAVIS (2.2.6).</p>
<p>For whole transcriptome sequencing, reads are aligned using STAR (2.7.3a) and expressions quantified using RSEM (1.3.3). Fusions are called using STAR-Fusion (1.8.1) and Arriba (1.2.0), followed by post processing with MAVIS (2.2.6). Variant prioritization is based on OncoKB actionable tiers (described below). In addition, variants without an OncoKB actionable tier are reported if they are listed as oncogenic, likely oncogenic or predicted oncogenic. Based on a minimum tumour purity of 30%, the sensitivity for SNVs and INDELs is 96% and 89%, respectively. The sensitivity for CNVs and FNA fusions is 86% and 32%, respectively. The limit of detection is 10% VAF for SNVs and 20% for INDELs.</p>
<p><strong>Genes tested</strong></p>
<p>Although whole genome and transcriptome sequencing encompasses all genes in a specimen, this report is restricted to cancer genes as defined by OncoKB as of the date this report.</p>
