
<!--This file is the HTML code for the Copy Number Variation (CNV) section of the report.-->


<!-- IMPORTS -->
<%
  from djerba.plugins.wgts.snv_indel_tools.render import html_builder
  import djerba.plugins.wgts.snv_indel_tools.constants as constants
%>
  

<!-- TAR SNV Indels -->
${html_builder().section_cells_begin("<h2>SNVs and in/dels</h2>","main")}
   
      <p>
            <strong>${html_builder().k_comma_format(results.get(constants.TOTAL_VARIANTS))}</strong> somatic mutation(s) were detected in exonic or splice regions, 
            of which <strong>${results.get(constants.TMB_TOTAL)}</strong> impacted a coding sequence,
            and <strong>${results.get(constants.CLINICALLY_RELEVANT_VARIANTS)}</strong> corresponded to an oncogenic mutation, as defined by OncoKB.
      </p>
      
      <img id='VAF' style='width: 100%; object-fit: contain' src="${results.get(constants.VAF_PLOT)}"/>

      % if results.get(constants.CLINICALLY_RELEVANT_VARIANTS) > 0:
            <table class="variants" width="100%">
                  ${html_builder().oncogenic_small_mutations_and_indels_header(results)}
                  <tbody>
                  % for row in html_builder().oncogenic_small_mutations_and_indels_rows(results):
                        ${row}
                  % endfor
                  </tbody>
            </table>
	    <table class="suppl" width="100%">
              <tr>
		<td width="50%" vertical-align="top"><b>Chr.</b>: Chromosome and cytoband</td>
		% if results.get(constants.HAS_EXPRESSION_DATA):
		<td width="50%" vertical-align="top"><b>Expr. (%)</b>: Expression Percentile for gene mRNA, or NA if comparison data is not available</td></tr>
	        % endif
	    </table>
      % endif

      
${html_builder().section_cells_end()}
