<%
from djerba.util.html import html_builder
import djerba.plugins.hrd.constants as hrd
from djerba.plugins.genomic_landscape.render import html_builder as gl_html_builder
%>

${html_builder.section_cells_begin("HRD", True)}

      <p>Homologous Recombination Deficiency (HRD) is the loss of a key DNA damage repair pathway in cancer, 
            canonically associated with loss-of-function mutations in <em>BRCA1/2</em>. HRD manifests as a 
            composite genomic signature that includes 
            somatic mutations (SNVs and in/dels), Loss-Of-Heterozygosity (LOH), copy number alterations 
            and other structural variants, summed as an HRD score. This sample shows 
            signatures consistent with <strong>${results.get(hrd.HRD_long)}</strong>.
      </p>

      <table class="variants" style="width:100%">
            <thead>
                  <th style=" width:10%">Biomarker</th>
                  <th style=" width:10%">Call</th>
                  <th style=" width:80%">Score & Confidence</th>
            </thead>
            <tbody>
                  <tr style="text-align:left; background-color:white"><td>HRD</td><td>${results.get(hrd.HRD_short)}</td><td><img id="HRD" style="width: 100%;" src="${results.get(hrd.HRD_PLOT)}"/></td>
            </tbody>


      </table>

      <p>
            Hrdetect takes SNVs and In/Dels from mutect2. The proportion of deletions that are at microhomologous sites is summarized as "Microhomologous Deletions".
            The counts of SNVs are categorized into exposures based on their trinucleotide context using 
            <a href="https://cran.r-project.org/web/packages/deconstructSigs/index.html">DeconstructSigs</a> (v. 1.8.0) and SBS signatures
            as defined in <a href="https://cancer.sanger.ac.uk/signatures/downloads/">COSMIC version 1</a>. 
            HRDetect also takes in LOH and structural variants. Structural variants are first called by 
            <a href="https://github.com/PapenfussLab/gridss">GRIDSS</a> (v.2.13.2)
            and then passed to PURPLE (v.3.8.1) for integrated LOH calling. Structural variants are then categorized into exposures based on break-end 
            characteristics using <a href="https://github.com/Nik-Zainal-Group/signature.tools.lib">signature.tools.lib</a> (v. 2.1.2) 
            and the rearrangement signature set defined in <a href="https://www.nature.com/articles/nature17676">Nik-Zainal et al. (2016)</a>. 
            For this test, sensitivity is 83% and specificity of 90%. The lower limit of detection has not been determined.
      </p>

      
${html_builder.section_cells_end()}
