<p>This test was developed and its performance characteristics determined by OICR Genomics. It has not been cleared or approved by the US Food and Drug Administration.</p>
<p> This assay combines two comprehensive next generation sequencing assays: a DNA-based whole genome sequencing (WGS) assay and an RNA-based whole transcriptome sequencing (WTS) assay. WGS libraries are prepared using the KAPA Hyper Prep kit with DNA extracted from FFPE or fresh frozen tissue (for tumour samples) or buffy coat blood specimens (for matched normal blood samples). Paired-end sequencing is performed using the NovaSeq6000 technology. For whole genome sequencing, alignments are performed using <a href="https://bio-bwa.sourceforge.net/">bwa</a> mem (0.7.12) against reference genome <a href="https://www.ncbi.nlm.nih.gov/assembly/GCF_000001405.39/">GRCh38.p13</a> and processed according to GATK best practices and includes duplicate marking with <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037052812-MarkDuplicates-Picard-">Picard</a> (2.21.2), indel realignment, and base recalibration with <a href="https://gatk.broadinstitute.org/hc/en-us/sections/360008481611-4-1-6-0">GATK</a> (4.1.6.0). SNVs and INDELs are called using <a href="https://gatk.broadinstitute.org/hc/en-us/articles/5358911630107-Mutect2">MuTect2</a> (GATK 4.2.6.1) and annotated with <a href="https://useast.ensembl.org/info/docs/tools/vep/index.html">VariantEffectPredictor</a> (105.0) followed by OncoKB, using <a href="https://www.ncbi.nlm.nih.gov/refseq/MANE/">MANE</a> transcripts (version 1.0). Copy number variations are called using <a href="https://doi.org/10.1093/annonc/mdu479">Sequenza</a> (2.1.2). Structural alterations are called using Delly (0.8.1) and post-processed using MAVIS (2.2.6).</p>
<p>For whole transcriptome sequencing, reads are aligned using STAR (2.7.3a) and expressions quantified using RSEM (1.3.3). Fusions are called using STAR-Fusion (1.8.1) and Arriba (1.2.0), followed by post processing with MAVIS (2.2.6). Variant prioritization is based on OncoKB actionable tiers (described below). In addition, variants without an OncoKB actionable tier are reported if they are listed as oncogenic, likely oncogenic or predicted oncogenic. Based on a minimum tumour purity of 30%, the sensitivity for SNVs and INDELs is 96% and 89%, respectively. The sensitivity for CNVs and FNA fusions is 86% and 32%, respectively. The limit of detection is 10% VAF for SNVs and 20% for INDELs.</p>
<p>Assay results are collated into the report document by <a href="https://github.com/oicr-gsi/djerba">Djerba</a> (${djerba_version}).</p>
<p><strong>Genes tested</strong></p>
<p>Although whole genome and transcriptome sequencing encompasses all genes in a specimen, this report is restricted to cancer genes as defined by OncoKB as of the date this report.</p>