<%
  import djerba.render.constants as constants
%>

<p>
    Whole Genome libraries were prepared using the KAPA Hyper Prep kit with DNA extracted from FFPE or fresh frozen tissue (for tumour samples) 
    or buffy coat blood specimens (for matched normal blood samples). Paired-end sequencing was performed using the Illumina NovaSeq6000 technology. 
    Reads were aligned using <a href=${versions.get(constants.BWAMEM_LINK)}>bwa</a> mem (${versions.get(constants.BWAMEM_VERSION)}) against
    reference genome <a href=${versions.get(constants.REFERENCE_GENOME_LINK)}>${versions.get(constants.REFERENCE_GENOME_VERSION)}</a> and processed according to GATK best practices, 
    including duplicate marking with <a href=${versions.get(constants.PICARD_LINK)}>Picard</a> (${versions.get(constants.PICARD_VERSION)}), 
    realignment around small insertions and deletions (in/dels), and base recalibration with <a href=${versions.get(constants.GATK_LINK)}>GATK</a> (${versions.get(constants.GATK_VERSION)}). 
    SNVs and in/dels were called using <a href=${versions.get(constants.MUTECT2_LINK)}>MuTect2</a> (GATK ${versions.get(constants.MUTECT2_VERSION)}) 
    and annotated with <a href=${versions.get(constants.VARIANTEFFECTPREDICTOR_LINK)}>VariantEffectPredictor</a> (${versions.get(constants.VARIANTEFFECTPREDICTOR_VERSION)}) using MANE  transcripts (MANE Clinical version 1.0 when available, <a href=${versions.get(constants.MANE_LINK)}>MANE Select</a> version 1.0 for all other transcripts).
    Variants were further annotated for oncogenicity and actionability by <a href="https://api.oncokb.org/oncokb-annotator/gene-annotation">OncoKB</a>. 
    In cases where OncoKB does not use MANE Select, links in annotation use the corresponding alteration in OncoKB.
    Copy number variations were called using <a href=${versions.get(constants.SEQUENZA_LINK)}>Sequenza</a> (${versions.get(constants.SEQUENZA_VERSION)}). 
</p>
