<p>
    Whole Genome libraries are prepared using the KAPA Hyper Prep kit with DNA extracted from FFPE or fresh frozen tissue (for tumour samples) 
    or buffy coat blood specimens (for matched normal blood samples). Paired-end sequencing is performed using the NovaSeq6000 technology. 
    Alignments are performed using <a href="https://bio-bwa.sourceforge.net/">bwa</a> mem (0.7.12) against 
    reference genome <a href="https://www.ncbi.nlm.nih.gov/assembly/GCF_000001405.39/">GRCh38.p13</a> and processed according to GATK best practices, 
    and include duplicate marking with <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037052812-MarkDuplicates-Picard-">Picard</a> (2.21.2), 
    indel realignment, and base recalibration with <a href="https://gatk.broadinstitute.org/hc/en-us/sections/360008481611-4-1-6-0">GATK</a> (4.1.6.0). 
    SNVs and INDELs are called using <a href="https://gatk.broadinstitute.org/hc/en-us/articles/5358911630107-Mutect2">MuTect2</a> (GATK 4.2.6.1) 
    and annotated with <a href="https://useast.ensembl.org/info/docs/tools/vep/index.html">VariantEffectPredictor</a> (105.0) followed by OncoKB, 
    using <a href="https://www.ncbi.nlm.nih.gov/refseq/MANE/">MANE</a> transcripts (version 1.0). 
    Copy number variations are called using <a href="https://doi.org/10.1093/annonc/mdu479">Sequenza</a> (2.1.2). 
    Microsatellite (MS) Instability status is called using <a href="https://www.sciencedirect.com/science/article/pii/S1672022920300218">msisensor-pro</a> (1.2.0) 
    and a custom list of MS sites created by msisensor-pro for hg38. 
</p>