
<!--This file is the HTML code for the Copy Number Variation (CNV) section of the report.-->


<!-- IMPORTS -->
<%
import djerba.plugins.wgts.cnv_purple.constants as constants
from djerba.util.image_to_base64 import converter
from djerba.plugins.wgts.cnv_purple.json_to_html import html_builder
%>
  

<!-- Copy Number Variants -->
${html_builder().section_cells_begin("<h2>CNV PURPLE DEV</h2>","main")}

      <p>The percent genome altered (PGA) was <strong>${results.get(constants.PERCENT_GENOME_ALTERED)}</strong>%. 
            <strong>${results.get(constants.TOTAL_VARIANTS)}</strong> cancer gene(s) were subject to copy number variation, 
            of which <strong>${results.get(constants.CLINICALLY_RELEVANT_VARIANTS)}</strong> corresponded to an oncogenic alteration, as defined by OncoKB. 
	    Regions with large copy number gains (&#8805; 6 CN) marked as &#9650 in plot below.
      </p>

      <img id='CNV' style='width: 100%; object-fit: contain' src="${results.get('cnv_plot')}"/>

      % if results.get(constants.CLINICALLY_RELEVANT_VARIANTS) > 0:
            <table class="variants" width="100%">
            ${html_builder().oncogenic_CNVs_header(results)}
            <tbody>
            % for row in html_builder().oncogenic_CNVs_rows(results):
                  ${row}
            % endfor

            </tbody>
            </table>



      % endif
      
${html_builder().section_cells_end()}